internal/pipeline/pipeline.go

```diff
@@ -164,3 +197,85 @@ type Function struct {
 	// `ConfigMap` is a convenient way to specify a function config of kind ConfigMap.
 	ConfigMap map[string]string `yaml:"configMap,omitempty"`
 }
+
+// Validate will validate the function `f`.
+func (f *Function) Validate() error {
+	err := ValidateFunctionName(f.Image)
+	if err != nil {
+		return fmt.Errorf("function name is not valid: %w", err)
+	}
+	// configCnt is used to count the number of configs provided. More than one
+	// config is not allowed.
+	var configCnt = 0
+	if f.ConfigPath != "" {
+		configCnt++
+		if err := ValidatePath(f.ConfigPath); err != nil {
+			return fmt.Errorf("configPath is invalid: %w", err)
+		}
+	}
+	if len(f.ConfigMap) != 0 {
+		configCnt++
+	}
+	if !isNodeZero(&f.Config) {
+		configCnt++
+	}
+	if configCnt > 1 {
+		return fmt.Errorf("only zero or one config is allowed, %d provided", configCnt)
+	}
+
+	return nil
+}
+
+// ValidateFunctionName validates the function name.
+// According to Docker implementation
+// https://github.com/docker/distribution/blob/master/reference/reference.go. A valid
+// name definition is:
+//	name                            := [domain '/'] path-component ['/' path-component]*
+//	domain                          := domain-component ['.' domain-component]* [':' port-number]
+//	domain-component                := /([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])/
+//	port-number                     := /[0-9]+/
+//	path-component                  := alpha-numeric [separator alpha-numeric]*
+// 	alpha-numeric                   := /[a-z0-9]+/
+//	separator                       := /[_.]|__|[-]*/
+func ValidateFunctionName(name string) error {
+	pathComponentRegexp := `(?:[a-z0-9](?:(?:[_.]|__|[-]*)[a-z0-9]+)*)`
+	domainComponentRegexp := `(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])`
+	domainRegexp := fmt.Sprintf(`%s(?:\.%s)*(?:\:[0-9]+)?`, domainComponentRegexp, domainComponentRegexp)
+	nameRegexp := fmt.Sprintf(`^(?:%s\/)?%s(?:\/%s)*$`, domainRegexp,
+		pathComponentRegexp, pathComponentRegexp)
+	matched, err := regexp.MatchString(nameRegexp, name)
+	if err != nil {
+		return err
+	}
+	if !matched {
+		return fmt.Errorf("function name %s isn't valid", name)
+	}
+	return nil
+}
+
+// IsNodeZero returns true if all the public fields in the Node are empty.
+// Which means it's not initialized and should be omitted when marshal.
+// The Node itself has a method IsZero but it is not released
+// in yaml.v3. https://pkg.go.dev/gopkg.in/yaml.v3#Node.IsZero
+// TODO: Use `IsYNodeZero` method from kyaml when kyaml has been updated to
+// >= 0.10.5
+func isNodeZero(n *yaml.Node) bool {
+	return n != nil && n.Kind == 0 && n.Style == 0 && n.Tag == "" && n.Value == "" &&
+		n.Anchor == "" && n.Alias == nil && n.Content == nil &&
+		n.HeadComment == "" && n.LineComment == "" && n.FootComment == "" &&
+		n.Line == 0 && n.Column == 0
+}
+
+// ValidatePath validates `path` and return an error if it's invalid
+func ValidatePath(p string) error {
```

Comment:
Do we need to export these (both Validate) functions ?

Reply:
Exposing these functions makes it can be tested in `pipeline_test` package. I am ok to use `pipeline` package in `pipeline_test.go`. Which one do you think is better?

Reply:
I think it is fine in this case because we don't expect lot of refactoring in this part of the code. 

Testing internal function can introduce maintenance headache if the code has a potential for lot of refactoring because now the maintainer has to not only change test code whenever they refactor internal func (and it's usage). So need to be done with careful judgement.